src/wp-admin/includes/file.php

```diff
@@ -684,7 +684,21 @@ function wp_tempnam( $filename = '', $dir = '' ) {
 	// Suffix some random data to avoid filename conflicts.
 	$temp_filename .= '-' . wp_generate_password( 6, false );
 	$temp_filename .= '.tmp';
-	$temp_filename  = $dir . wp_unique_filename( $dir, $temp_filename );
+	$temp_filename  = wp_unique_filename( $dir, $temp_filename );
+
+	/*
+	 * Filesystems typically have a limit of 255 characters for a filename.
+	 *
+	 * If the generated unique filename exceeds this, truncate the initial
+	 * filename and try again.
+	 */
+	$characters_over_255 = strlen( $temp_filename ) - 255;
+	if ( $characters_over_255 > 0 ) {
+		$filename = substr( $filename, 0, -$characters_over_255 );
```

Comment:
Thinking this should shorten the name to less than 255, 252 or even 250 should be good. Reason is that if a file with the shortened name already exists, `wp_unique_filename()` (from the second run of `wp_tempnam()`) will append a dash and numbers to it to make it unique bringing it over 255 again.

Also seems this can be written as:
```
if ( strlen( $temp_filename ) > 255 ) {
    $filename = substr( $filename, 0, 250 );
    return wp_tempnam( $filename, $dir );
}
```

Reply:
As the check occurs after `wp_unique_filename()`, I don't see the point of failure you're referring to.

Given an original `$filename` of ` (254) ...very_long_name`:

1. Add `(11) -123456.tmp`: ` (265) ...very_long_name-123456.tmp`
2. If it exists, add `(2) -1`: ` (267) ...very_long_name-123456-1.tmp`
3. If it exceeds 255 characters, subtract the overflow (13) from the **original** filename: `(241) ...v`
4. Repeat steps 1-3 on `(241) ...v`:
5. Add `(11) -123456.tmp`: ` (252) ...v-123456.tmp`
6. If it exists, add `(2) -1`: `(string) (254) ...v-123456-1.jpg`
7. Does not exceed 255 characters :white_tick:

Even if `...v-123456.tmp` through `...v-123456-9.tmp` existed and `...v-123456-10.tmp` was generated, this would exceed the limit and be truncated before running again. The likelihood of this happening at all is very low, as `123456` is a random 6-character password. If it did occur, the chances of it happening on the next run are "don't even leave the house!" levels of bad luck.

Adding a hard 250 means each update to the function will need to examine whether the 250 needs to change. The implementation above makes it a fluid "whatever exceeds 255", and should be futureproof as long as the 255 doesn't change on the filesystems :sweat_smile:

Reply:
I suspect that wp_unique_filename() needs to be called after cutting the file name and not after.

Reply:
`wp_unique_filename()` is called when `wp_tempnam()` is called again after cutting the original filename. If we only called `wp_unique_filename()` again after cutting, then the random data to avoid filename conflicts wouldn't be added, increasing the likelihood of conflicts.

We could store the random data and just append it to the cut filename before re-running `wp_unique_filename()`. However, by not running `wp_tempnam()` again, it remains possible that `(255) ... cut-filename-randomdata.jpg` exists, and the `wp_unique_filename()` call generates `(257) ...cut-filename-randomdata-1.jpg`, producing the error this aims to solve.

Instead, by cutting the filename and re-running `wp_tempnam()`, the intention is to guarantee that a filename over 255 characters can never be produced.

Reply:
Hmmmm, still thinking I'm missing something? :) 

> Given an original $filename of  (254) ...very_long_name:
> 
> Add (11) `-123456.tmp`:  (265) ...very_long_name-123456.tmp
> If it exists, add (2) `-1`:  (267) ...very_long_name-123456-1.tmp

Lets assume it doesn't exist. At this point the filename is 265 chars long.

> If it exceeds 255 characters, subtract the overflow (13) from the original filename: (241) ...v

The overflow is 10 now, 254 (the original length) + 11 - 255.

> Repeat steps 1-3 on (241) ...v

On 244 now, 254 (the original length) - 10.

> Add (11) -123456.tmp:  (252) ...v-123456.tmp

Brings it to 255.

> If it exists, add (2) -1: (string) (254) ...v-123456-1.jpg

Brings it to 257. 

Another variation: in the original example if there are 9 matching files already (yea, extremely unlikely...), `wp_unique_filename()` from the new call to `wp_tempnam()` will add `'-10'` instead of `'-1'` bringing the total to 256?

So still thinking that "cutting off" 3 more chars would make this work smoother? I.e. instead of 
`$characters_over_255 = strlen( $temp_filename ) - 255;`
can do 
`$characters_over_255 = strlen( $temp_filename ) - 252;`

Reply:
Ah, so this isn't about a point of failure, but rather about trying to reduce the potential number of recursions? I can make the change, though I'll rename variables and add an inline comment to explain the 252 as it may seem strange or arbitrary otherwise.

Reply:
Updated in [0f60029](https://github.com/WordPress/wordpress-develop/pull/4647/commits/0f600290dfa8a9935e9fb08d2013f63bf618ae22).